twisted/protocols/test/test_tls.py

```diff
@@ -1522,3 +1523,219 @@ def test_interface(self):
+            after the TLS handshake is complete.
+        @type clientProtocols: L{list} of L{bytes}
+
+        @param serverProtocols: The protocols the server will accept speaking
+            after the TLS handshake is complete.
+        @type serverProtocols: L{list} of L{bytes}
+
+        @return: A L{tuple} of four different items: the client L{Protocol},
+            the server L{Protocol}, a L{Deferred} that fires when the client
+            first receives bytes (and so the TLS connection is complete), and a
+            L{Deferred} that fires when the server first receives bytes.
+        @rtype: A L{tuple} of (L{Protocol}, L{Protocol}, L{Deferred},
+            L{Deferred})
+        """
+        bytes = b'some bytes'
+
+        class NotifyingSender(Protocol):
+            def __init__(self, notifier):
+                self.notifier = notifier
+
+            def connectionMade(self):
+                self.transport.writeSequence(list(iterbytes(bytes)))
+
+            def dataReceived(self, bytes):
+                if self.notifier is not None:
+                    self.notifier.callback(self)
+                    self.notifier = None
+
+
+        clientDataReceived = Deferred()
+        clientFactory = ClientNegotiationFactory(clientProtocols)
+        clientFactory.protocol = lambda: NotifyingSender(
+            clientDataReceived
+        )
+
+        clientContextFactory, _ = (
+            HandshakeCallbackContextFactory.factoryAndDeferred())
+        wrapperFactory = TLSMemoryBIOFactory(
+            clientContextFactory, True, clientFactory)
+        sslClientProtocol = wrapperFactory.buildProtocol(None)
+
+        serverDataReceived = Deferred()
+        serverFactory = ServerNegotiationFactory(serverProtocols)
+        serverFactory.protocol = lambda: NotifyingSender(
+            serverDataReceived
+        )
+
+        serverContextFactory = ServerTLSContext()
+        wrapperFactory = TLSMemoryBIOFactory(
+            serverContextFactory, False, serverFactory)
+        sslServerProtocol = wrapperFactory.buildProtocol(None)
+
+        loopbackAsync(
+            sslServerProtocol, sslClientProtocol
+        )
+        return (sslClientProtocol, sslServerProtocol, clientDataReceived,
+                serverDataReceived)
+
+
+    def test_negotiationWithNoProtocols(self):
+        """
+        When factories support L{IProtocolNegotiationFactory} but don't
+        advertise support for any protocols, no protocols are negotiated.
+        """
+        client, server, clientDataReceived, serverDataReceived = (
+            self.handshakeProtocols([], [])
+        )
+
+        def checkNegotiatedProtocol(ignored):
+            self.assertEqual(client.negotiatedProtocol, None)
+            self.assertEqual(server.negotiatedProtocol, None)
+
+        clientDataReceived.addCallback(lambda ignored: serverDataReceived)
+        serverDataReceived.addCallback(checkNegotiatedProtocol)
```

Comment:
Maybe instead of `serverDataReceived.addCallback(checkNegotiatedProtocol)`
we should have `clientDataReceived.addCallback(checkNegotiatedProtocol)`

as otherwise the assertions are called before the client is connected.